sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java

```diff
@@ -0,0 +1,335 @@
+  // ================================================================================
+
+  /**
+   * Which factory to use for creating Pubsub transport.
+   */
+  private final PubsubClientFactory pubsubFactory;
+
+  /**
+   * Pubsub topic to publish to.
+   */
+  private final PubsubClient.TopicPath topic;
+
+  /**
+   * Coder for elements. It is the responsibility of the underlying Pubsub transport to
+   * re-encode element bytes if necessary, eg as Base64 strings.
+   */
+  private final Coder<T> elementCoder;
+
+  /**
+   * Pubsub metadata field holding timestamp of each element, or {@literal null} if should use
+   * Pubsub message publish timestamp instead.
+   */
+  @Nullable
+  private final String timestampLabel;
+
+  /**
+   * Pubsub metadata field holding id for each element, or {@literal null} if need to generate
+   * a unique id ourselves.
+   */
+  @Nullable
+  private final String idLabel;
+
+  /**
+   * Number of 'shards' to use so that latency in Pubsub publish can be hidden. Generally this
+   * should be a small multiple of the number of available cores. Too smoll a number results
+   * in too much time lost to blocking Pubsub calls. To large a number results in too many
+   * single-element batches being sent to Pubsub with high per-batch overhead.
+   */
+  private final int numShards;
+
+  public PubsubUnboundedSink(
+      PubsubClientFactory pubsubFactory,
+      PubsubClient.TopicPath topic,
+      Coder<T> elementCoder,
+      String timestampLabel,
+      String idLabel,
+      int numShards) {
+    this.pubsubFactory = pubsubFactory;
+    this.topic = topic;
+    this.elementCoder = elementCoder;
+    this.timestampLabel = timestampLabel;
+    this.idLabel = idLabel;
+    this.numShards = numShards;
+  }
+
+  public PubsubClient.TopicPath getTopic() {
+    return topic;
+  }
+
+  @Nullable
+  public String getTimestampLabel() {
+    return timestampLabel;
+  }
+
+  @Nullable
+  public String getIdLabel() {
+    return idLabel;
+  }
+
+  public Coder<T> getElementCoder() {
+    return elementCoder;
+  }
+
+  @Override
+  public PDone apply(PCollection<T> input) {
+    // TODO: Include topic.getPath() in transform metadata when it is available.
+    input.apply(
+        Window.named("PubsubSink.Window")
+            .<T>into(new GlobalWindows())
+            .triggering(
+                Repeatedly.forever(
+                    AfterFirst.of(AfterPane.elementCountAtLeast(PUBLISH_BATCH_SIZE),
+                                  AfterProcessingTime.pastFirstElementInPane()
+                                                     .plusDelayOf(MAX_LATENCY))))
+            .discardingFiredPanes())
+         .apply(ParDo.named("PubsubSink.Shard").of(new ShardFn<T>(elementCoder, numShards)))
+         .setCoder(KvCoder.of(VarIntCoder.of(), CODER))
+         .apply(GroupByKey.<Integer, PubsubClient.OutgoingMessage>create())
+         .apply(ParDo.named("PubsubSink.Writer").of(new WriterFn(pubsubFactory, topic,
```

Comment:
I like that it's part of the transform, just like all the other properties. Then apply is just a verbose pipe.